feat: add merge to setNodes and test

zhugexinxin opened the pull request:
**Description**
I create a new PR to help you adjust.

**Checks**
- [x] The new code matches the existing patterns and styles.
- [x] The tests pass with `yarn test`.
- [x] The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
- [x] The relevant examples still work. (Run examples with `yarn start`.)
- [ ] You've [added a changeset](https://github.com/atlassian/changesets/blob/master/docs/adding-a-changeset.md) if changing functionality. (Add one with `yarn changeset add`.)

zhugexinxin commented:
> **Description** I create a new PR to help you adjust.
> 
> **Checks**
> 
> * [x ] The new code matches the existing patterns and styles.
> * [x]  The tests pass with `yarn test`.
> * [x]  The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
> * [x ] The relevant examples still work. (Run examples with `yarn start`.)
> * [ ]  You've [added a changeset](https://github.com/atlassian/changesets/blob/master/docs/adding-a-changeset.md) if changing functionality. (Add one with `yarn changeset add`.)

@dylans 

is this what you expect?

zhugexinxin commented:
> > **Description** I create a new PR to help you adjust.
> > **Checks**
> > 
> > * [x]  The new code matches the existing patterns and styles.
> > * [x]   The tests pass with `yarn test`.
> > * [x]   The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
> > * [x]  The relevant examples still work. (Run examples with `yarn start`.)
> > * [ ]   You've [added a changeset](https://github.com/atlassian/changesets/blob/master/docs/adding-a-changeset.md) if changing functionality. (Add one with `yarn changeset add`.)
> 
> @dylans
> 
> is this what you expect? sorry, I'm so sorry, I didn't understand what you meant at first, I deleted the branch when I saw the merge. It seems that I can only roll back the previous merge and merge the current one, I am very sorry for this

dylans commented:
> > **Description** I create a new PR to help you adjust.
> > **Checks**
> > 
> > * [x]  The new code matches the existing patterns and styles.
> > * [x]   The tests pass with `yarn test`.
> > * [x]   The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
> > * [x]  The relevant examples still work. (Run examples with `yarn start`.)
> > * [ ]   You've [added a changeset](https://github.com/atlassian/changesets/blob/master/docs/adding-a-changeset.md) if changing functionality. (Add one with `yarn changeset add`.)
> 
> @dylans
> 
> is this what you expect? I'm so sorry, I didn't understand what you meant at first, I deleted the branch when I saw the merge. It seems that I can only roll back the previous merge and merge the current one, I am very sorry for this

I've addressed it... when you raise a PR in the future please just either run `yarn changeset` or follow the prompts in the PR to create one. This helps automate the release notes for each release. Thanks!

zhugexinxin commented:
> I've addressed it... when you raise a PR in the future please just either run `yarn changeset` or follow the prompts in the PR to create one. This helps automate the release notes for each release. Thanks!

Ok, I get it now

zhugexinxin commented:
> > > **Description** I create a new PR to help you adjust.
> > > **Checks**
> > > 
> > > * [x]   The new code matches the existing patterns and styles.
> > > * [x]    The tests pass with `yarn test`.
> > > * [x]    The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
> > > * [x]   The relevant examples still work. (Run examples with `yarn start`.)
> > > * [ ]    You've [added a changeset](https://github.com/atlassian/changesets/blob/master/docs/adding-a-changeset.md) if changing functionality. (Add one with `yarn changeset add`.)
> > 
> > 
> > @dylans
> > is this what you expect? I'm so sorry, I didn't understand what you meant at first, I deleted the branch when I saw the merge. It seems that I can only roll back the previous merge and merge the current one, I am very sorry for this
> 
> I've addressed it... when you raise a PR in the future please just either run `yarn changeset` or follow the prompts in the PR to create one. This helps automate the release notes for each release. Thanks!



> > > **Description** I create a new PR to help you adjust.
> > > **Checks**
> > > 
> > > * [x]   The new code matches the existing patterns and styles.
> > > * [x]    The tests pass with `yarn test`.
> > > * [x]    The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
> > > * [x]   The relevant examples still work. (Run examples with `yarn start`.)
> > > * [ ]    You've [added a changeset](https://github.com/atlassian/changesets/blob/master/docs/adding-a-changeset.md) if changing functionality. (Add one with `yarn changeset add`.)
> > 
> > 
> > @dylans
> > is this what you expect? I'm so sorry, I didn't understand what you meant at first, I deleted the branch when I saw the merge. It seems that I can only roll back the previous merge and merge the current one, I am very sorry for this
> 
> I've addressed it... when you raise a PR in the future please just either run `yarn changeset` or follow the prompts in the PR to create one. This helps automate the release notes for each release. Thanks!

Thank you very much, the change and the release will be very important to me.

zhugexinxin commented:
@dylans 
Will the robot release the version automatically?

dylans commented:
> @dylans
> Will the robot release the version automatically?

I will do a release tonight. I like to release when I know I’ll be around for a few hours in case anything goes wrong.

bkrausz commented on packages/slate/src/interfaces/editor.ts:
```diff
@@ -1736,3 +1736,4 @@ export type NodeMatch<T extends Node> =
   | ((node: Node, path: Path) => boolean)
 
 export type PropsCompare = (prop: Partial<Node>, node: Partial<Node>) => boolean
+export type PropsMerge = (prop: Partial<Node>, node: Partial<Node>) => object
```
If I'm not mistaken, isn't this TS def (and the one for Compare) incorrect? It should be expecting a single property rather than a partial object of properties. And it won't necessarily get an object back, it could get anything.

A slightly more accurate version would be
```ts
type PropsMerge<T = unknown> = (prop: T | undefined, node: T) => T
```

bkrausz replied:
`PropsCompare` has the same bug FWIW.

bkrausz commented:
I'm curious if we're happy with this API, or if it was put in place to fix a single issue (and if so, should we generalize it). It doesn't, for example, tell you the specific key being modified (neither does `compare`).

IMO a clearer API would be a single method that gives you the matched node and expects a list of properties to set. i.e.:
```ts
type PropsGenerator = (node: Node) => Partial<Node>
setNodes<T extends Node>(
  editor: Editor,
  props: Partial<Node> | PropsGenerator,
  options: {
    at?: Location
    match?: NodeMatch<T>
    mode?: 'all' | 'highest' | 'lowest'
    hanging?: boolean
    split?: boolean
    voids?: boolean
    compare?: PropsCompare
  } = {}
): void;
```

Then if someone wanted to do a deep compare & deep update, they'd do something like:

```ts
setNodes(editor, (node) => {
  if (!isEqual(node.key, newValue)) {
    return { key: deepCopy(newValue) };
  } else {
    return {};
  }
}, { match });
```

It's a bit more verbose on the client side, but it's:
1. more explicit
2. IMO clearer: the second param is either props or a function that returns props based on the node
3. Enabled other use cases. E.g., I stumbled onto this change because I want `setNodes` to set a unique ID for each node it transforms. I could accomplish this via:
```ts
setNodes(editor, () => ({type: 'foo', id: generateId()}), { match });
```

I'm happy to produce a PR to that effect, but wanted to confirm that this is reasonable before I do so. @dylans please lmk if you have any thoughts/concerns with this approach.
